ARCHITECTURE.md

```diff
@@ -0,0 +1,88 @@
+# Cargo Architecture
+
+This document gives a high level overview of Cargo internals. You may
+find it useful if you want to contribute to Cargo or if you are
+interested in the inner workings of Cargo.
+
+
+## Subcommands
+
+Cargo is organized as a set of subcommands. All subcommands live in
+`src/bin` directory. However, only `src/bin/cargo.rs` file produces an
+executable, other files inside the `bin` directory are submodules. See
+`src/bin/cargo.rs` for how these subcommands get wired up with the
```

Comment:
Yeah, it was a bit shameful to explain ourselves here :) I think the layout should be

```
src/
  bin/
    cargo/
      main.ru
      buld.rs
      test.rs
```

to match that of https://github.com/rust-lang/cargo/issues/4086. I would like to fix #4086 first though, to make sure we follow conventional layout.